### PR TITLE
Integrate finetuning & fingerprinting

### PR DESCRIPTION
**Pull request by WenkelF**
## Changelogs

- Upgraded finetuning to allow finetuning on custom datasets
- Integrated fingerprinting
- Added example datasets for custom finetuning/fingerprinting
- Adding documentation for finetuning/fingerprinting

---

_Checklist:_

- [ ] _Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR._
- [ ] _Add tests to cover the fixed bug(s) or the new introduced feature(s) (if appropriate)._
- [x] _Update the API documentation is a new function is added, or an existing one is deleted._
- [x] _Write concise and explanatory changelogs above._
- [x] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

### CODE REVIEW

File: expts/data/finetuning_example-cls/raw.csv
Comment (DomInvivo) [inline expts/data/finetuning_example-cls/raw.csv:1]: Please crop the file. You could have 200 compounds for the sake of the example.

File: expts/data/finetuning_example-cls/split.csv
Comment (DomInvivo) [inline expts/data/finetuning_example-cls/split.csv:1]: Also crop the file according to the previous comment

File: expts/data/finetuning_example-cls/raw.csv
Comment (DomInvivo) [reply expts/data/finetuning_example-cls/raw.csv:1]: Alternatively, check if there are other example data that could be used as testing file

File: expts/data/finetuning_example-reg/raw.csv
Comment (DomInvivo) [inline expts/data/finetuning_example-reg/raw.csv:1]: Also crop to 200 compounds for the sake of the example

File: expts/data/finetuning_example-reg/split.csv
Comment (DomInvivo) [inline expts/data/finetuning_example-reg/split.csv:1]: crop accordingly

File: expts/hydra-configs/finetuning/custom.yaml
Comment (DomInvivo) [inline expts/hydra-configs/finetuning/custom.yaml:1]: Better name than `custom.yaml`? Perhaps `example.yaml` if it's the one used for the examples? or `default_finetuning.yaml` if it's the default one always used for finetuning?

File: expts/hydra-configs/finetuning/custom.yaml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+finetuning:
+  task: finetune
+  level: graph
+  pretrained_model: dummy-pretrained-model
+  finetuning_module: graph_output_nn  
+  sub_module_from_pretrained: graph
+  new_sub_module: graph
+  drop_depth: 1
+  added_depth: 1
+  new_out_dim: 1024 # TODO: infer automatically
```
Comment (DomInvivo) [inline expts/hydra-configs/finetuning/custom.yaml:75]: Are you finetuning on 1024 tasks at the same time? Or is that the hidden dim?

Perhaps replacing the `TODO` by `# Change according to the number of finetuning tasks`

File: expts/hydra-configs/fingerprinting/custom.yaml
Comment (DomInvivo) [inline expts/hydra-configs/fingerprinting/custom.yaml:1]: Same comment about file name

File: graphium/finetuning/finetuning.py [hunk 2]
```diff
@@ -29,7 +29,8 @@ def __init__(
         finetuning_module: str,
         added_depth: int = 0,
         unfreeze_pretrained_depth: Optional[int] = None,
-        epoch_unfreeze_all: int = 0,
+        epoch_unfreeze_all: Optional[int] = 0,
+        freeze_always: Optional[Union[List, str]] = None,
```
Comment (DomInvivo) [inline graphium/finetuning/finetuning.py:33]: `freeze_always` -> `always_freeze_modules`

Also docstring is missing for that variable, It think it's important

File: graphium/fingerprinting/data.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+from graphium.data.datamodule import MultitaskFromSmilesDataModule, ADMETBenchmarkDataModule
+from graphium.trainer.predictor import PredictorModule
+from graphium.fingerprinting.fingerprinter import Fingerprinter
+
+
+class FingerprintDataset(Dataset):
+    def __init__(
+            self,
+            smiles: List[str],
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:22]: Shouldn't the Smiles be the last item, and be optional? Since they're not returned, unless it's for debugging or testing purposes?
And if `smiles is not None`, than also return the smiles in the `__getitem__`? Otherwise there's no point in having the smiles

File: graphium/fingerprinting/data.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+from pytorch_lightning import LightningDataModule
+
+from torch.utils.data import Dataset, DataLoader
+
+from graphium.data.datamodule import MultitaskFromSmilesDataModule, ADMETBenchmarkDataModule
+from graphium.trainer.predictor import PredictorModule
+from graphium.fingerprinting.fingerprinter import Fingerprinter
+
+
+class FingerprintDataset(Dataset):
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:19]: Missing docs for parameters of that class.

File: graphium/fingerprinting/data.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        return len(self.smiles)
+
+    def __getitem__(self, index):
+        fp_list = []
+        for val in self.fingerprints.values():
+            fp_list.append(val[index])
+        return fp_list, self.labels[index]
+
+
+class FingerprintDatamodule(LightningDataModule):
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:40]: Missing docs for parameters of that class. You can just copy/paste from the MultiTaskDatamodule

Comment (DomInvivo) [inline graphium/fingerprinting/data.py:40]: Why not inherit from the MultitaskDatamodule? Is it because you sometimes need the "ADMETDatamodule"?

File: graphium/fingerprinting/data.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            for model, layers in self.pretrained_models.items():
+                predictor = PredictorModule.load_pretrained_model(model, device=self.device)
+                predictor.featurization.pop("max_num_atoms", None)
+
+                # Featurization
+                if self.benchmark is None:
+                    assert self.df_path is not None, "df_path must be provided if not using an integrated benchmark"
+
+                    # Add a dummy task column (filled with NaN values) in case no such column is provided
+                    smiles_df = pd.read_csv(self.df_path)
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:123]: Too restrictive. Can't finetune on parquet or other? Use the `MultitaskDataModule.read_table`?

File: graphium/fingerprinting/data.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                    task_cols = [col for col in smiles_df if col.startswith("task_")]
+                    if len(task_cols) == 0:
+                        df_path = ".".join(self.df_path.split(".")[:-1])
+                        smiles_df["task_dummy"] = np.nan
+                        smiles_df.to_csv(f"{df_path}_with_dummy_task_col.csv", index=False)
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:128]: Why saving the smiles to CSV?

File: graphium/fingerprinting/data.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+                            "smiles_col": "smiles",
+                            "label_cols": "task_*",
+                            "task_level": "graph",
+                            "splits_path": self.splits_path,
+                            "split_type": self.split_type,
+                            "split_val": self.split_val,
+                            "split_test": self.split_test,
+                            "seed": self.data_seed,
+                        }
+                    }
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:143]: A lot of shared logic with the class `TaskSpecificArgs` I think

File: graphium/fingerprinting/data.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                    datamodule = MultitaskFromSmilesDataModule(
+                        task_specific_args=task_specific_args,
+                        batch_size_inference=128,
+                        featurization=predictor.featurization,
+                        featurization_n_jobs=0,
+                        processed_graph_data_path=f"{self.mol_cache_dir}/mols/",
+                    )
+
+                elif self.benchmark == "tdc":
+                    datamodule = ADMETBenchmarkDataModule(
```
Comment (DomInvivo) [inline graphium/fingerprinting/data.py:155]: Perhaps change the name of that class to `TDCBenchmarkDatamodule`? Or is it used for Polaris too? Then `PolarisTDCBenchmarkDatamodule`?

File: graphium/fingerprinting/fingerprinter.py [hunk 10]
```diff
@@ -152,32 +153,61 @@ def get_fingerprints_for_batch(self, batch):
...
+    def get_fingerprints_for_dataset(self, dataloader, store_dict: bool=False):
         """Return the fingerprints for an entire dataset"""
 
         original_out_type = self._out_type
         self._out_type = "torch"
 
         fps = []
         for batch in tqdm.tqdm(dataloader, desc="Fingerprinting batches"):
-            feats = self.get_fingerprints_for_batch(batch)
+            feats = self.get_fingerprints_for_batch(batch, store_dict=store_dict)
```
Comment (DomInvivo) [inline graphium/fingerprinting/fingerprinter.py:188]: How does that work for large datasets? Do you store everything in memory?

**Review by DomInvivo — CHANGES_REQUESTED** [review]
Left a few comments, but great work in general :)

File: expts/data/finetuning_example-cls/raw.csv
Comment (WenkelF) [reply expts/data/finetuning_example-cls/raw.csv:1]: Cropped to 200 compounds [here](https://github.com/datamol-io/graphium/pull/529/commits/807913f616066136d013f4c91ab5f58cf2de6f83).

File: expts/data/finetuning_example-reg/raw.csv
Comment (WenkelF) [reply expts/data/finetuning_example-reg/raw.csv:1]: This one only has 160 compounds already.

File: graphium/fingerprinting/data.py (hunk 7 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py:128]: This is to allow users to provide compounds without an associated task column (which is expected by the datamodule). We can save it in the same format as the original file.

File: expts/hydra-configs/finetuning/custom.yaml
Comment (WenkelF) [reply expts/hydra-configs/finetuning/custom.yaml:1]: Fixed [here](https://github.com/datamol-io/graphium/pull/529/commits/697e3d1012ab244caf164418b0ebf50407ce00fc).

File: expts/hydra-configs/fingerprinting/custom.yaml
Comment (WenkelF) [reply expts/hydra-configs/fingerprinting/custom.yaml:1]: Fixed [here](https://github.com/datamol-io/graphium/pull/529/commits/697e3d1012ab244caf164418b0ebf50407ce00fc).

File: graphium/fingerprinting/data.py (hunk 3 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py]: Agree and made it optional [here](https://github.com/datamol-io/graphium/pull/529/commits/d9aa407d774934e170e8d8af7028155b676b5981).

File: graphium/fingerprinting/data.py (hunk 4 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py:19]: Fixed [here](https://github.com/datamol-io/graphium/pull/529/commits/d9aa407d774934e170e8d8af7028155b676b5981#diff-a660520d63c09919ce7b7d4450b3edf709bba2c5d21fa421a52fa4d1fff7a9e3).

File: graphium/fingerprinting/data.py (hunk 5 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py:52]: Also fixed [here](https://github.com/datamol-io/graphium/pull/529/commits/d9aa407d774934e170e8d8af7028155b676b5981#diff-a660520d63c09919ce7b7d4450b3edf709bba2c5d21fa421a52fa4d1fff7a9e3).

Comment (WenkelF) [reply graphium/fingerprinting/data.py:52]: Yes, `MultitaskDatamodule` or `TDCDatamodule` are used depending on the data source. There could be an argument for making `FingerprintDatamodule.prepare_data` a function (`create_fps`) of the `BaseDataModule` but this would require major changes. I suggest keeping it in mind for future improvements.

File: graphium/fingerprinting/data.py (hunk 6 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py]: Using `MultitaskDataModule.read_table` now as suggested (see [here](https://github.com/datamol-io/graphium/pull/529/commits/d9aa407d774934e170e8d8af7028155b676b5981#diff-c2ceb6f93c9d4be9d916f0840dbd40f0af5be3566a18d13fd2857f0c20b537e3)).

File: graphium/fingerprinting/data.py (hunk 8 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py]: Now using `DatasetProcessingParams`.

File: graphium/fingerprinting/data.py (hunk 9 above)
Comment (WenkelF) [reply graphium/fingerprinting/data.py]: Changed to `TDCBenchmarkDatamodule`.

**Review by DomInvivo — APPROVED** [review]
